README.md
## Update for Rails 5.1+

I've started to use the [premailer](https://github.com/premailer/premailer) gem for Rails > 5.0. I'd welcome patches to keep this gem going.

## Inline Styles Mailer [![Build Status](https://secure.travis-ci.org/billhorsman/inline_styles_mailer.png)](http://travis-ci.org/billhorsman/inline_styles_mailer)

Using Jack Danger's excellent [Inline Styles](https://github.com/jackdanger/inline_styles) gem is even easier if you're using Rails 3.1+ and this gem.

The Inline Styles gem helps you embed CSS styles into your markup so that you can send pretty HTML emails that won't get butchered by email clients that strip out CSS. Or, more precisely, will help reduce the amount of butchering (even with inline CSS some styles, like background images, are often cut out).

This gem stands on the shoulders of the [inline_styles](https://github.com/jackdanger/inline_styles) and [sass-rails](https://github.com/rails/sass-rails) gem, merely adding some code to make it more convenient to use.

## What do you mean, "inline style"?

Let's say you have a mail template like this:

```html
<html>
  <body>
    <p>Hello World</p>
  </body>
</html>
```

And a CSS file like this:

```css
p {
  color: red;
}
```

Then this gem will mash that all up into:

```html
<html>
  <body>
    <p style="color: red;">Hello World</p>
  </body>
</html>
```

So you can keep your templates clean and take advantage of CSS preprocessing goodness if you want without compromising the portability of your HTML in various email clients.

MailChimp have a nice article: [How To Code HTML Emails](http://kb.mailchimp.com/article/how-to-code-html-emails/).

## Installation

If you're using Bundler:

```ruby
source 'http://rubygems.org'
gem 'inline_styles_mailer'
```

## Usage

If you keep things simple, then it's just one line:

```ruby
class FooMailer < ActionMailer::Base
  include InlineStylesMailer

  def foo(email)
    mail(:to => email, :subject => "Foo foo!")
  end

end
```

If you have a CSS file <code>app/assets/stylesheets/_foo_mailer*</code> (where * can be .css, .css.scss or .css.sass) then it will get automatically applied to the mail using the inline_styles gem. That name (<code>_foo_mailer</code>) is based on your mailer class name, e.g. FooMailer. If you have more than one file matching that pattern then it will use them all.

It will use one of three preprocessing methods based on the filename:

* [.scss](http://sass-lang.com/)
* [.sass](http://sass-lang.com/)
* anything else (e.g. .css) no preprocessing at all

Want to use a different css file? Declare <code>use_stylesheet</code>:

```ruby
class FooMailer < ActionMailer::Base
  include InlineStylesMailer
  use_stylesheet '_bar.css.sass'
  ...
end
```

You can use an array of stylesheets if you like. Don't keep your stylesheets in <code>app/assets/stylesheets</code>? Declare <code>stylesheet_path</code>:

```ruby
class FooMailer < ActionMailer::Base
  include InlineStylesMailer
  stylesheet_path 'public/stylesheets'
  ...
end
```

## Upgrading to Version 1.0

We decided to bump up the major version because of a fix (5d0b6f4) we made to the way the template was found. We were incorrectly assuming that all mail templates were only nested one deep in `app/views` but that's not always the case. It now works correctly but you might have worked around this bug (issue #8) by moving your templates. You don't need to do that any more.

## What versions of Rails does this work with?

We run tests using Rails `4.2.6`, `5.0.0.beta3`, `5.1.6` and `5.2`

| Version | Working? |
| --- | --- |
| 3.0 | No * |
| 3.1 | Yes |
| 3.2 | Yes |
| 4.0 | Yes |
| 4.1 | Yes |
| 4.2 | Yes |
| 5.0 | Yes |
| 5.1 | Yes |
| 5.2 | Yes |

\* It relies on the asset pipeline so I'd be surprised if it works. Not tested.

\** I'm recommending using the [premailer](https://github.com/premailer/premailer) gem for Rails > 5.0.

## What about Ruby?

Anything later than Ruby `1.9.2` should be fine. We test using Ruby 2.0.0`, `2.1.8`, `2.2.4` and `2.3.0`.

### 1.9 Support

We don't test with Ruby `1.9` because there is a dependency on the `mime-types-data` gem for which the latest version requires Ruby `2.0` or later. That doesn't mean you can't use Ruby `1.9.3` but you'll need to make sure you're using a version of `mime-types-data` less than `3`.

## Development

Questions or problems? Please post them on the [issue tracker](https://github.com/billhorsman/inline_styles_mailer/issues). You can contribute changes by forking the project and submitting a pull request. You can ensure the tests passing by running `bundle` and `rake`.

The tests also run on [Travis CI](http://travis-ci.org/#!/billhorsman/inline_styles_mailer).

This gem was created by Bill Horsman and is under the MIT License.


CHANGELOG.md
## Next

## v1.0.3

Added support for Rails 5.2.1 (now simply defined as < 6)

## v1.0.2

Added support for Rails 5.1 and 5.2.

## v1.0.1

Now works with Rails 5.0 final (just a change to the `.gemspec`)

## v1.0.0

Changes to the way we find nested templates which might break the way you define their location (but probably won't affect most people). See the [note here](https://github.com/billhorsman/inline_styles_mailer#upgrading-to-version-10).

## v0.0.9

Now needs Ruby 1.9.3 (drop support for 1.9.2), because Rails 4.1 does so.
Works with Rails 4.2

## v0.0.8

Works with Rails 4.1

## v0.0.7

Honour correct order of email parts.

## v0.0.6

Smarter lookup of appropriate view template.

## v0.0.5

Drop support for Ruby 1.8.7.

## v0.0.4

Smarter location of CSS files and more explicit override if you need it.

## v0.0.3

Now copes with .sass and .css files (as well as .scss).

## v0.0.2

Documentation changes only.

## v0.0.1

Initial release


spec/foo_mailer.rb
require 'rails/all'

class FooMailer < ActionMailer::Base
  include InlineStylesMailer

  def foo
    mail(:to => "test@localhost", :subject => "Test")
  end

  def backwards
    mail(:to => "test@localhost", :subject => "Test", :parts_order => [ "text/html", "text/plain"])
  end

end

spec/spec_helper.rb
require 'inline_styles_mailer'
require 'rails/all'

ActionMailer::Base.delivery_method = :test
ActionMailer::Base.prepend_view_path File.join(File.dirname(__FILE__), "fixtures", "views")

Dir["#{File.dirname(__FILE__)}/fixtures/mailers/**/*.rb"].each { |f| require f }

RSpec.configure do |config|
  config.expect_with :rspec do |c|
    c.syntax = :should
  end
  config.mock_with :rspec do |c|
    c.syntax = :should
  end
end


spec/fixtures/mailers/foo/bar_mailer.rb
require 'rails/all'

module Foo
  class BarMailer < ActionMailer::Base
    include InlineStylesMailer

    def bar
      mail(:to => "test@localhost", :subject => "Test Bar")
    end
  end
end

lib/inline_styles_mailer.rb
require "inline_styles"
require "inline_styles_mailer/version"
require "sass"

module InlineStylesMailer

  def self.included(base)
    base.extend(ClassMethods)
  end

  module ClassMethods

    def use_stylesheet(*stylesheets)
      @stylesheets = Array(stylesheets)
    end

    def stylesheet_path(path)
      @stylesheet_path = path
    end

    def locate_template(name)
      "#{self.name.underscore}/#{name}.html"
    end

    def css_content
      @stylesheet_path ||= File.join("app", "assets", "stylesheets")
      @css_content ||= stylesheets.map {|stylesheet|
        Dir[Rails.root.join(@stylesheet_path, "#{stylesheet}")].map {|file|
          case file
          when /\.scss$/
            Sass::Engine.new(File.read(file), syntax: :scss).render
          when /\.sass$/
            Sass::Engine.new(File.read(file), syntax: :sass).render
          else
            # Plain old CSS? Let's assume it is.
            File.read(file)
          end
        }.join("\n")
      }.compact.join("\n")
      @css_content
    end

    def stylesheets
      @stylesheets ||= [self.name.underscore.sub(/^(.*\/)*(.*)$/, '\1_\2*')]
    end

    def page
      @page ||= InlineStyles::Page.new.with_css(css_content)
    end

    # For testing
    def reset
      @css_content = nil
      @page = nil
    end

  end

  def mail(options, &block)
    if block
      super(mail, &block) # We'll just let this pass through
    else
      super(options) do |format|
        # Rails 3.1 takes an array, while Rails 3.0 takes a string.
        # See https://github.com/billhorsman/inline_styles_mailer/issues/1
        prefixes = options[:template_path] || self.class.name.underscore
        prefixes = [prefixes].flatten unless Rails.version =~ /^3\.0/
        templates = lookup_context.find_all(options[:template_name] || action_name, prefixes)
        options.reverse_merge!(:mime_version => "1.0", :charset => "UTF-8", :content_type => "text/plain", :parts_order => [ "text/plain", "text/enriched", "text/html"])
        templates.sort_by {|t|
          # Rails 4.1 use #type but earlier versions use #mime_type
          mime_type = t.respond_to?(:mime_type) ? t.mime_type : t.type
          i = options[:parts_order].index(mime_type)
          i > -1 ? i : 99
        }.each do |template|
        # templates.each do |template|
          # e.g. template = app/views/user_mailer/welcome.html.erb
          # e.g. template = app/views/namespace/user_mailer/welcome.html.erb
          template_path = template.inspect.split("views")[1][1..-1] # e.g. user_mailer/welcome.html.erb
          parts = template_path.split('.')
          handler = parts.pop.to_sym # e.g. erb
          extension = parts.pop.to_sym # e.g. html
          file = parts.join('.') # e.g. user_mailer/welcome (you get a deprecation warning if you don't strip off the format and handler)
          format.send(extension) do
            case extension
            when :html
              html = render_to_string :file => file, :layout => layout_to_use, handlers: [handler]
              # Rails 5.1 removed render :text
              if Gem.loaded_specs['rails'].version >= Gem::Version.create('5.0')
                render :plain => self.class.page.with_html(html).apply
              else
                render :text => self.class.page.with_html(html).apply
              end
            else
              render
            end
          end
        end
      end
    end
  end

  def layout_to_use
    case call_layout
    when ActionView::Template
      call_layout.inspect.split("/").last.split(".").first
    when String
      call_layout.split("/").last.split(".").first
    end
  end

  # Hack to call _layout the right way depending on the Rails version. This is a code smell
  # telling us that we shouldn't be doing this at all...
  def call_layout
    if method(:_layout).arity == 1
      # Rails 5?
      _layout([:html])
    else
      # < Rails 5?
      _layout
    end
  end

end
